File: src/opt/passes/copy-propagation.ts
import type { MirInstruction, MirPlace, MirTerminator, MirValue } from "../../mir/mir.js";
import type { MirPass } from "../mir-pass.js";

export const copyPropagationPass: MirPass = {
  name: "copy-propagation",
  run(module) {
    let changed = false;

    for (const func of module.functions) {
      for (const block of func.blocks) {
        const copies = new Map<string, MirValue>();

        for (const instruction of block.instructions) {
          changed = rewriteInstruction(instruction, copies) || changed;

          if (instruction.kind === "call" || instruction.kind === "store") {
            copies.clear();
            continue;
          }

          const target = getInstructionTarget(instruction);
          if (target?.kind === "temp") {
            copies.delete(target.name);
          }

          if (instruction.kind === "move" && instruction.target.kind === "temp") {
            copies.set(instruction.target.name, instruction.value);
          }
        }

        changed = rewriteTerminator(block.terminator, copies) || changed;
      }
    }

    return { changed };
  }
};

function rewriteInstruction(instruction: MirInstruction, copies: Map<string, MirValue>): boolean {
  let changed = false;
  switch (instruction.kind) {
    case "const_int":
    case "const_float":
    case "const_bool":
      return false;
    case "move": {
      const value = resolveCopy(instruction.value, copies);
      if (value !== instruction.value) {
        instruction.value = value;
        changed = true;
      }
      return changed;
    }
    case "binary": {
      const left = resolveCopy(instruction.left, copies);
      const right = resolveCopy(instruction.right, copies);
      if (left !== instruction.left) {
        instruction.left = left;
        changed = true;
      }
      if (right !== instruction.right) {
        instruction.right = right;
        changed = true;
      }
      return changed;
    }
    case "unary": {
      const operand = resolveCopy(instruction.operand, copies);
      if (operand !== instruction.operand) {
        instruction.operand = operand;
        changed = true;
      }
      return changed;
    }
    case "compare": {
      const left = resolveCopy(instruction.left, copies);
      const right = resolveCopy(instruction.right, copies);
      if (left !== instruction.left) {
        instruction.left = left;
        changed = true;
      }
      if (right !== instruction.right) {
        instruction.right = right;
        changed = true;
      }
      return changed;
    }
    case "address":
      return rewritePlace(instruction.place, copies);
    case "load":
      return rewritePlace(instruction.place, copies);
    case "store": {
      changed = rewritePlace(instruction.place, copies);
      const value = resolveCopy(instruction.value, copies);
      if (value !== instruction.value) {
        instruction.value = value;
        changed = true;
      }
      return changed;
    }
    case "call": {
      for (let index = 0; index < instruction.args.length; index += 1) {
        const arg = instruction.args[index]!;
        const value = resolveCopy(arg, copies);
        if (value !== arg) {
          instruction.args[index] = value;
          changed = true;
        }
      }
      return changed;
    }
  }
}

function rewriteTerminator(terminator: MirTerminator, copies: Map<string, MirValue>): boolean {
  switch (terminator.kind) {
    case "return": {
      const value = resolveCopy(terminator.value, copies);
      if (value !== terminator.value) {
        terminator.value = value;
        return true;
      }
      return false;
    }
    case "branch": {
      const condition = resolveCopy(terminator.condition, copies);
      if (condition !== terminator.condition) {
        terminator.condition = condition;
        return true;
      }
      return false;
    }
    case "jump":
      return false;
  }
}

function rewritePlace(place: MirPlace, copies: Map<string, MirValue>): boolean {
  switch (place.kind) {
    case "param":
    case "local":
      return false;
    case "deref": {
      const pointer = resolveCopy(place.pointer, copies);
      if (pointer !== place.pointer) {
        place.pointer = pointer;
        return true;
      }
      return false;
    }
    case "field":
      return rewritePlace(place.base, copies);
    case "index": {
      let changed = rewritePlace(place.base, copies);
      const index = resolveCopy(place.index, copies);
      if (index !== place.index) {
        place.index = index;
        changed = true;
      }
      return changed;
    }
  }
}

function resolveCopy(value: MirValue, copies: Map<string, MirValue>): MirValue {
  let current = value;
  const seen = new Set<string>();

  while (current.kind === "temp") {
    if (seen.has(current.name)) {
      return current;
    }
    seen.add(current.name);
    const next = copies.get(current.name);
    if (!next) {
      return current;
    }
    current = next;
  }

  return current;
}

function getInstructionTarget(instruction: MirInstruction): MirValue | undefined {
  switch (instruction.kind) {
    case "const_int":
    case "const_float":
    case "const_bool":
    case "move":
    case "binary":
    case "unary":
    case "compare":
    case "address":
    case "load":
    case "call":
      return instruction.target;
    case "store":
      return undefined;
  }
}

File: src/optimizer/passes/copy_propagation.rs
use std::collections::{HashMap, HashSet};

use crate::{MirInstruction, MirModule, MirPlace, MirTerminator, MirValue, instruction_effect};

use super::super::{analysis::*, pipeline::*};

pub(in crate::optimizer) fn run_copy_propagation(
    module: &mut MirModule,
    _context: &MirPassContext,
) -> MirPassResult {
    let mut changed = false;
    for function in &mut module.functions {
        for block in &mut function.blocks {
            let mut copies = HashMap::new();
            for instruction in &mut block.instructions {
                changed |= rewrite_instruction_copies(instruction, &copies);

                if instruction_effect(instruction).invalidates_value_facts() {
                    copies.clear();
                    continue;
                }

                if let Some(target) = instruction_target(instruction)
                    && let Some(name) = temp_name(target)
                {
                    copies.remove(name);
                }

                if let MirInstruction::Move { target, value } = instruction
                    && let Some(name) = temp_name(target)
                {
                    copies.insert(name.to_string(), value.clone());
                }
            }
            changed |= rewrite_terminator_copies(&mut block.terminator, &copies);
        }
    }

    MirPassResult {
        changed,
        diagnostics: Vec::new(),
    }
}

fn rewrite_instruction_copies(
    instruction: &mut MirInstruction,
    copies: &HashMap<String, MirValue>,
) -> bool {
    let mut changed = false;
    match instruction {
        MirInstruction::ConstInt { .. }
        | MirInstruction::ConstFloat { .. }
        | MirInstruction::ConstBool { .. } => false,
        MirInstruction::Move { value, .. } | MirInstruction::Cast { value, .. } => {
            changed |= rewrite_value_copy(value, copies);
            changed
        }
        MirInstruction::Binary { left, right, .. }
        | MirInstruction::Compare { left, right, .. } => {
            changed |= rewrite_value_copy(left, copies);
            changed |= rewrite_value_copy(right, copies);
            changed
        }
        MirInstruction::Unary { operand, .. } => {
            changed |= rewrite_value_copy(operand, copies);
            changed
        }
        MirInstruction::MakeSlice { data, len, .. } => {
            changed |= rewrite_value_copy(data, copies);
            changed |= rewrite_value_copy(len, copies);
            changed
        }
        MirInstruction::SliceData { slice, .. } | MirInstruction::SliceLen { slice, .. } => {
            rewrite_value_copy(slice, copies)
        }
        MirInstruction::Subslice {
            slice, start, end, ..
        } => {
            changed |= rewrite_value_copy(slice, copies);
            changed |= rewrite_value_copy(start, copies);
            changed |= rewrite_value_copy(end, copies);
            changed
        }
        MirInstruction::Address { place, .. } | MirInstruction::Load { place, .. } => {
            rewrite_place_copies(place, copies)
        }
        MirInstruction::Store { place, value } => {
            changed |= rewrite_place_copies(place, copies);
            changed |= rewrite_value_copy(value, copies);
            changed
        }
        MirInstruction::Call { args, .. } | MirInstruction::RuntimeCall { args, .. } => {
            for arg in args {
                changed |= rewrite_value_copy(arg, copies);
            }
            changed
        }
    }
}

fn rewrite_terminator_copies(
    terminator: &mut MirTerminator,
    copies: &HashMap<String, MirValue>,
) -> bool {
    match terminator {
        MirTerminator::Return { value } => value
            .as_mut()
            .is_some_and(|value| rewrite_value_copy(value, copies)),
        MirTerminator::Branch { condition, .. } => rewrite_value_copy(condition, copies),
        MirTerminator::Jump { .. } => false,
    }
}

fn rewrite_place_copies(place: &mut MirPlace, copies: &HashMap<String, MirValue>) -> bool {
    match place {
        MirPlace::Param { .. } | MirPlace::Local { .. } => false,
        MirPlace::Deref { pointer, .. } => rewrite_value_copy(pointer, copies),
        MirPlace::Field { base, .. } => rewrite_place_copies(base, copies),
        MirPlace::Index { base, index, .. } => {
            rewrite_place_copies(base, copies) | rewrite_value_copy(index, copies)
        }
        MirPlace::SliceIndex { slice, index, .. } => {
            rewrite_value_copy(slice, copies) | rewrite_value_copy(index, copies)
        }
    }
}

fn rewrite_value_copy(value: &mut MirValue, copies: &HashMap<String, MirValue>) -> bool {
    let resolved = resolve_copy(value, copies);
    if *value == resolved {
        return false;
    }
    *value = resolved;
    true
}

fn resolve_copy(value: &MirValue, copies: &HashMap<String, MirValue>) -> MirValue {
    let mut current = value.clone();
    let mut seen = HashSet::new();
    while let MirValue::Temp { name, .. } = &current {
        if !seen.insert(name.clone()) {
            return current;
        }
        let Some(next) = copies.get(name) else {
            return current;
        };
        current = next.clone();
    }
    current
}
